### PR TITLE
Ease log and output code maintenance

### LINKED ISSUE #475: Improve on logging and stderr'ing stuff
#### Is your feature request related to a problem?

Not really, more like a maintenance issue.

#### Describe the feature you'd like

I'd like there to be log functions (for both the log file and `stderr`, and similar) so we wouldn't keep appending to a file here and there, for example. This surfaced as an idea during the implementation of #471.

#### Describe alternatives you've considered

Not changing the code 😄

### PR DESCRIPTION
**Pull request by paulo-ferraz-oliveira**
# Description

Closes #475.

Initially opening as draft, since I still wanna test this better.

**Edit**: I've done the complete self-review, at this moment, but I'm still gonna give it a second pass by replaying all changes using the patch diff.

- [x] I have performed a self-review of my changes
- [x] I have read and understood the [contributing guidelines](/kerl/kerl/blob/main/CONTRIBUTING.md)

### CODE REVIEW

**Comment by paulo-ferraz-oliveira** [comment]
(this one doesn't have to/shouldn't go into the next release)

File: kerl [hunk 1]
```diff
@@ -79,19 +160,19 @@ fi
...
-    "notice" | "note" | "n")
+    "notice")
         ret="$(tput setaf "${KERL_COLOR_N:=4}")"
         ;;
-    "tip" | "t")
+    "tip")
         ret="$(tput setaf "${KERL_COLOR_T:=6}")"
         ;;
-    "success" | "s")
+    "success")
```
Comment (paulo-ferraz-oliveira) [inline kerl:175]: Simplified for readability.

File: kerl [hunk 2]
```diff
@@ -62,10 +62,92 @@ stderr() {
...
+usage() {
+    # $1: usage tip
+
+    l=error stderr "usage: $0 $1"
+}
+
+error() {
+    # $1: log entry
+
+    l=error stderr "ERROR: $1"
```
Comment (paulo-ferraz-oliveira) [inline kerl:80]: Prefix for consistency.

File: kerl [hunk 3]
```diff
@@ -62,10 +62,92 @@ stderr() {
...
+}
+
+warn() {
+    # $1: log entry
+
+    l=warn stderr "WARNING: $1"
```
Comment (paulo-ferraz-oliveira) [inline kerl:86]: Prefix for consistency.

File: kerl [hunk 4]
```diff
@@ -62,10 +62,92 @@ stderr() {
...
+    l=success stderr "$1"
+}
+
+tip() {
+    # $1: log entry
+
+    l=tip stderr "$1"
+}
+
+init_build_logfile() {
```
Comment (paulo-ferraz-oliveira) [inline kerl:107]: A log init function, to encapsulate behaviour.

File: kerl [hunk 5]
```diff
@@ -62,10 +62,92 @@ stderr() {
...
+}
+
+log_build_cmd() {
+    # $*: command
+    build_cmd="$*"
+
+    eval "$build_cmd" >>"$BUILD_LOGFILE" 2>&1
+}
+
+init_install_logfile() {
```
Comment (paulo-ferraz-oliveira) [inline kerl:130]: Another log init function, to encapsulate behaviour.

**Comment by paulo-ferraz-oliveira** [comment]
Asking for review (and moving away from draft) but this can be merged after the release so we have time to fix any issues that might arise (mostly it's logging, but there's some eval there too, so I'm taking CI seriously too 😄)

File: kerl [hunk 6]
```diff
@@ -62,10 +62,92 @@ stderr() {
...
 }
 
+nocolor() {
+    # $1: log entry
+
+    stderr "$1"
+}
+
+usage() {
+    # $1: usage tip
```
Comment (jadeallenx) [inline kerl:72]: what's $0 here?

File: kerl [hunk 7]
```diff
@@ -259,8 +341,8 @@ fi
 if [ "$KERL_BUILD_BACKEND" = 'git' ]; then
     KERL_USE_AUTOCONF=1
 elif [ "$KERL_BUILD_BACKEND" != 'tarball' ]; then
-    l=e stderr "Unhandled value KERL_BUILD_BACKEND='${KERL_BUILD_BACKEND}'"
-    l=t stderr "KERL_BUILD_BACKEND must be one of 'git' (default) or 'tarball'"
+    error "enhandled value KERL_BUILD_BACKEND=${KERL_BUILD_BACKEND}."
```
Comment (jadeallenx) [inline kerl:344]: Typo here

File: kerl (hunk 6 above)
Comment (paulo-ferraz-oliveira) [reply kerl:72]: The filename of the script that is currently being executed.
I could add variable names to everything not scoped to a function.
I'll do that.

File: kerl (hunk 7 above)
Comment (paulo-ferraz-oliveira) [reply kerl:344]: Nice catch. I'm gonna check this is CSpell.

Comment (paulo-ferraz-oliveira) [reply kerl]: 5f5f773

File: kerl (hunk 6 above)
Comment (paulo-ferraz-oliveira) [reply kerl:74]: 8d7e9ec

**Comment by paulo-ferraz-oliveira** [comment]
I'm gonna:

* make the `usage` function call `exit 1` because that's what's always being done
* move `_usage` function closer to their use (so there's no "jump" to check "documentation")
* move all usage exits to the top of their respective check spots

**Review by jadeallenx — COMMENTED** [review]
looks really good so far!

**Comment by paulo-ferraz-oliveira** [comment]
New commit, b185051, where I try to simplify input maintenance (and relation to `usage`).

**Comment by paulo-ferraz-oliveira** [comment]
New commit, b36fbf8, where I approach the implementation to current expectations...

**Comment by paulo-ferraz-oliveira** [comment]
https://github.com/kerl/kerl/actions/runs/6487803829/job/17619275468?pr=476#step:11:1 is failing without output. I know it's an issue with the code, and I've fixed it already, but I'm wondering if we can't output errors to non-terminal (which is what's happening).

**Comment by paulo-ferraz-oliveira** [comment]
@jadeallenx, I'm wondering if we should change `stderr` to include something like

```bash
    elif [ -n "$KERL_DEBUG" ] && [ -n "$CI" ]; then
        echo "$*"
```

The prior relevant changes were done in https://github.com/kerl/kerl/commit/76090303613dadbb2869a87c283d57b166f44a8f (#403).

**Comment by jadeallenx** [comment]
Getting output from CI is important so I'm +1 on that suggestion

**Comment by paulo-ferraz-oliveira** [comment]
I pushed that change. I don't mind (or rather I even prefer) if this gets merged after the release. Is that Ok with you?

**Comment by jadeallenx** [comment]
Yeah that's fine

**Comment by jadeallenx** [comment]
ready to merge here now?

**Comment by paulo-ferraz-oliveira** [comment]
Yes, lemme just check some things and I'll merge afterward.
There seems to be a "lint" issue in the main branch, but I can't replicate it 😄

**Comment by paulo-ferraz-oliveira** [comment]
Oh, I see what happened 😄. You changed some files pushing directly to the main branch, so CI didn't have a chance to complain. I don't mind fixing those, but would propose changes to go through CI, and not directly to the main branch.

**Pull request merged**